pkg/nodeagent/controller/operatingsystemconfig/containerd_config.go

```diff
@@ -0,0 +1,289 @@
+			importsPath:        {"imports"},
+			sandboxImagePath:   {"plugins", "io.containerd.grpc.v1.cri", "sandbox_image"},
+			cgroupDriverPath:   {"plugins", "io.containerd.grpc.v1.cri", "containerd", "runtimes", "runc", "options", "SystemdCgroup"},
+			cniPluginPath:      {"plugins", "io.containerd.grpc.v1.cri", "cni", "bin_dir"},
+		},
+		3: {
+			registryConfigPath: {"plugins", "io.containerd.cri.v1.images", "registry", "config_path"},
+			importsPath:        {"imports"},
+			sandboxImagePath:   {"plugins", "io.containerd.cri.v1.runtime", "sandbox_image"},
+			cgroupDriverPath:   {"plugins", "io.containerd.cri.v1.runtime", "containerd", "runtimes", "runc", "options", "SystemdCgroup"},
+			cniPluginPath:      {"plugins", "io.containerd.cri.v1.runtime", "cni", "bin_dir"},
+		},
+	}
+
+	// pluginPathReplacements is a map that contains replacements for paths brought in through an osc plugin config
+	pluginPathReplacements = replacementMap{
+		{"plugins", "io.containerd.grpc.v1.cri", "containerd", "runtimes"}: {"plugins", "io.containerd.cri.v1.runtime", "containerd", "runtimes"},
+	}
+)
+
+// getContainerdConfigFileVersion obtains the containerd configuration file version from the configuration file
+func getContainerdConfigFileVersion(config map[string]any) (containerdConfigFileVersion, error) {
+	version, ok := config["version"]
+	if !ok {
+		// config file versions 2 and 3 must contain a version header
+		// if it cannot be found, it therefore must be version 1
+		return 1, nil
+	}
+
+	i, ok := version.(int64)
+	if !ok {
+		return 0, fmt.Errorf("cannot assert containerd config file version as an int64")
+	}
+
+	if i > 3 {
+		return 0, fmt.Errorf("unsupported containerd config file version %d", i)
+	}
+
+	return containerdConfigFileVersion(i), nil
+}
+
+// ensureContainerdDefaultConfig invokes the 'containerd' and saves the resulting default configuration.
+func (r *Reconciler) ensureContainerdDefaultConfig(ctx context.Context) error {
+	exists, err := r.FS.Exists(configFile)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		return nil
+	}
+
+	output, err := Exec(ctx, "containerd", "config", "default")
+	if err != nil {
+		return err
+	}
+
+	return r.FS.WriteFile(configFile, output, 0644)
+}
+
+// ensureContainerdConfiguration sets the configuration for containerd.
+func (r *Reconciler) ensureContainerdConfiguration(log logr.Logger, criConfig *extensionsv1alpha1.CRIConfig) error {
+	config, err := r.FS.ReadFile(configFile)
+	if err != nil {
+		return fmt.Errorf("unable to read containerd config.toml: %w", err)
+	}
+
+	content := map[string]any{}
+
+	if err = toml.Unmarshal(config, &content); err != nil {
+		return fmt.Errorf("unable to decode containerd default config: %w", err)
+	}
+
+	containerdConfigFileVersion, err := getContainerdConfigFileVersion(content)
+	if err != nil {
+		return err
+	}
+
+	type (
+		patch struct {
+			name  string
+			path  structuredmap.Path
+			setFn structuredmap.SetFn
+		}
+	)
```

Comment:
Why the extra block?

```suggestion
	type patch struct {
		name  string
		path  structuredmap.Path
		setFn structuredmap.SetFn
	}
```